okhttp/src/main/java/io/grpc/okhttp/OkHttpClientStream.java

```diff
@@ -148,112 +139,163 @@ public void start(ClientStreamListener listener) {
-   */
-  @GuardedBy("lock")
-  public void transportDataReceived(okio.Buffer frame, boolean endOfStream) {
-    long length = frame.size();
-    window -= length;
-    if (window < 0) {
-      frameWriter.rstStream(id(), ErrorCode.FLOW_CONTROL_ERROR);
-      transport.finishStream(id(), Status.INTERNAL.withDescription(
-          "Received data size exceeded our receiving window size"), null);
-      return;
+  class TransportState extends Http2ClientStreamTransportState {
+    private final Object lock;
+    @GuardedBy("lock")
+    private List<Header> requestHeaders;
+    /**
+     * Null iff {@link #requestHeaders} is null.  Non-null iff neither {@link #sendCancel} nor
+     * {@link #start(int)} have been called.
+     */
+    @GuardedBy("lock")
+    private Queue<PendingData> pendingData = new ArrayDeque<PendingData>();
+    @GuardedBy("lock")
+    private boolean cancelSent = false;
+    @GuardedBy("lock")
+    private int window = Utils.DEFAULT_WINDOW_SIZE;
+    @GuardedBy("lock")
+    private int processedWindow = Utils.DEFAULT_WINDOW_SIZE;
+
+    public TransportState(int maxMessageSize, StatsTraceContext statsTraceCtx, Object lock) {
+      super(maxMessageSize, statsTraceCtx);
+      this.lock = checkNotNull(lock, "lock");
     }
-    super.transportDataReceived(new OkHttpReadableBuffer(frame), endOfStream);
-  }
 
-  @Override
-  protected void sendFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {
-    Buffer buffer;
-    if (frame == null) {
-      buffer = EMPTY_BUFFER;
-    } else {
-      buffer = ((OkHttpWritableBuffer) frame).buffer();
-      int size = (int) buffer.size();
-      if (size > 0) {
-        onSendingBytes(size);
+    @GuardedBy("lock")
+    public void start(int streamId) {
+      checkState(id == ABSENT_ID, "the stream has been started with id %s", streamId);
+      id = streamId;
+      state.onStreamAllocated();
+
+      if (pendingData != null) {
+        // Only happens when the stream has neither been started nor cancelled.
+        frameWriter.synStream(false, false, id, 0, requestHeaders);
+        statsTraceCtx.clientHeadersSent();
+        requestHeaders = null;
+
+        boolean flush = false;
+        while (!pendingData.isEmpty()) {
+          PendingData data = pendingData.poll();
+          outboundFlow.data(data.endOfStream, id, data.buffer, false);
+          if (data.flush) {
+            flush = true;
+          }
+        }
+        if (flush) {
+          outboundFlow.flush();
+        }
+        pendingData = null;
       }
     }
 
-    synchronized (lock) {
-      if (cancelSent) {
-        return;
-      }
-      if (pendingData != null) {
-        // Stream is pending start, queue the data.
-        pendingData.add(new PendingData(buffer, endOfStream, flush));
+    @GuardedBy("lock")
+    @Override
+    protected void onStreamAllocated() {
+      super.onStreamAllocated();
+    }
+
+    @GuardedBy("lock")
+    @Override
+    protected void onStreamDeallocated() {
```

Comment:
The problem is that transportReportStatus gets called in in two places:
1) AbstractClientStream2 line 252(headers not received before payload)
2) Http2ClientStreamTransportState line 156(trailer is not received)

Those two calls are not exactly from transport. The abstract layer finds out violation, reports the status and closes the stream.
And transport is not aware of these and don't know the stream gets closed. How does netty handle this?